webrender/src/device/gfx/descriptor.rs

```diff
@@ -2,409 +2,519 @@
-    ) -> Self {
-        let descriptor_pool = unsafe {
-            device.create_descriptor_pool(
-                max_size,
-                descriptor_range_descriptors.as_slice(),
-                hal::pso::DescriptorPoolCreateFlags::empty(),
-            )
+ impl Locals {
+    fn transform_as_u32_slice(&self) -> &[u32; 16] {
+        unsafe {
+            std::mem::transmute::<&[[f32; 4]; 4], &[u32; 16]>(&self.uTransform)
         }
-        .expect("create_descriptor_pool failed");
-        let descriptor_set_layout =
-            unsafe { device.create_descriptor_set_layout(&descriptor_set_layout, &[]) }
-                .expect("create_descriptor_set_layout failed");
-        let mut dp = DescPool {
-            descriptor_pool,
-            descriptor_set: vec![],
-            descriptor_set_layout,
-            current_descriptor_set_idx: 0,
-            max_descriptor_set_size: max_size,
-            free_sets: vec![]
-        };
-        dp.allocate();
-        dp
     }
+}
 
-    pub(super) fn make_descriptor_set(&mut self) -> (&B::DescriptorSet, usize) {
-        if let Some(idx) = self.free_sets.pop() {
-            return (&self.descriptor_set[idx], idx)
-        }
-        (&self.descriptor_set[self.current_descriptor_set_idx], self.current_descriptor_set_idx)
+ impl Hash for Locals {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.transform_as_u32_slice().hash(state);
+        self.uMode.hash(state);
+    }
+}
+
+ impl PartialEq for Locals {
+    fn eq(&self, other: &Locals) -> bool {
+        self.transform_as_u32_slice() == other.transform_as_u32_slice() &&
+        self.uMode == other.uMode
+    }
+}
+
+impl Eq for Locals {}
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug, Default)]
+pub(super) struct PerDrawBindings(pub [TextureId; PER_DRAW_TEXTURE_COUNT]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug, Default)]
+pub(super) struct PerPassBindings(pub [TextureId; PER_PASS_TEXTURE_COUNT]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug, Default)]
+pub(super) struct PerGroupBindings(pub [TextureId; PER_GROUP_TEXTURE_COUNT]);
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+pub(super) struct SamplerBindings(pub [TextureFilter; MUTABLE_SAMPLER_COUNT]);
+
+impl Default for SamplerBindings {
+    fn default() -> Self {
+        SamplerBindings([TextureFilter::Linear; MUTABLE_SAMPLER_COUNT])
+    }
+}
+
+pub(super) struct DescriptorGroupData<B: hal::Backend> {
+    pub(super) descriptor_set_layouts: FastHashMap<DescriptorGroup, ArrayVec<[B::DescriptorSetLayout; DESCRIPTOR_SET_PER_LAYOUT]>>,
+    pub(super) descriptor_set_ranges: FastHashMap<DescriptorGroup, ArrayVec<[DescriptorRanges; DESCRIPTOR_SET_PER_LAYOUT]>>,
+    pub(super) pipeline_layouts: FastHashMap<DescriptorGroup, B::PipelineLayout>,
+}
+
+impl<B: hal::Backend> DescriptorGroupData<B> {
+    pub(super) fn descriptor_layout(&self, group: &DescriptorGroup, group_idx: usize) -> &B::DescriptorSetLayout {
```

Comment:
correspondingly, you'd not need any of these accessor methods if you only had a single map